### PR TITLE
911 signin captcha

### PR DESCRIPTION
**Pull request by notmd**
Part of #911 
- Add Cloudflare captcha.

### CODE REVIEW

File: website/src/lib/captcha.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  if (action) {
+    if (cdata) {
+      return response.action === action && response.cdata === cdata;
+    }
+    return response.action === action;
+  }
+
+  return false;
+};
```
Comment (notmd) [inline website/src/lib/captcha.ts:56]: I'm unable to test this code cause with a demo key Cloudflare doesn't send action and cdata in response.

File: website/src/middleware.ts [hunk 2]
```diff
@@ -14,5 +16,38 @@ export const config = {
...
+    }
+
+    const url = req.nextUrl.clone();
+    url.pathname = "/api/auth/invalid-captcha";
+
+    return NextResponse.redirect(url);
+  }
+
+  return withAuth(req);
+};
```
Comment (notmd) [inline website/src/middleware.ts:39]: This is extremely hacky. I tried 2 other solutions but they did not work.
- `NextResponse.redirect('/auth/singin?errorCode=somecode')` cause next-auth `signIn` function doesn't follow redirect.
- `return NextResponse.json({url: '/auth/singin?errorCode=somecode'})` cause Nextjs doesn't allow to modify response body in middleware

File: .vscode/settings.json [hunk 3]
```diff
@@ -1,4 +1,5 @@
 {
   "python.formatting.provider": "black",
-  "python.analysis.extraPaths": ["${workspaceFolder}/oasst-shared"]
+  "python.analysis.extraPaths": ["${workspaceFolder}/oasst-shared"],
+  "prettier.singleQuote": false
```
Comment (yk) [inline .vscode/settings.json:4]: isn't this the default value?

Comment (notmd) [reply .vscode/settings.json:4]: Yes but my prettier config in the local is different. Without this change prettier will use single quote.

**Comment by notmd** [comment]
There is a small problem when the user fills in the email and clicks the sign-in button too fast, the captcha may not be loaded yet (it can take 2s depending on the internet connection). And the website will inform the captcha is invalid which may confuse the users because they don't see any captcha on the page. Here are a few solutions.
- Disable the button until the `onSuccess` callback gets called. But maybe there is a situation that the captcha never gets successful, and the user won't be able to log in.
- Check for the captcha response when the user clicks login and shows a toast.
- Switch to visible captcha.

File: website/src/lib/captcha.ts (hunk 1 above)
Comment (fozziethebeat) [reply website/src/lib/captcha.ts:56]: Can you add that as a comment?

File: website/src/middleware.ts (hunk 2 above)
Comment (fozziethebeat) [reply website/src/middleware.ts:39]: did you try using the signIn callback within NextAuth? If you modify `src/pages/api/auth/[...[nextauth].ts` You can fill out the `signIn` call back.  Returning true means signin works.  Returning false rejects the signin.  So I think you could put very similar code in that callback when the user is using the email provider.  Not entirely sure how yo get the captcha data tho.

**Review by fozziethebeat — CHANGES_REQUESTED** [review]
If you could, can you try the callback approach for captcha validation?

Also, for ensuring the captcha triggers I am a fan of
-  Disable email signin until captcha loads
- If loading fails for any reason, trigger a toast to tell the user to reload the page and try again

**Comment by yk** [comment]
maybe I'm just overlooking this right now, but is it possible that we have a setting to circumvent (i.e. disable) the captcha, maybe via an environment variable?

**Comment by notmd** [comment]
> maybe I'm just overlooking this right now, but is it possible that we have a setting to circumvent (i.e. disable) the captcha, maybe via an environment variable?

Sure, I will try to add it now.

**Review by 53jk1 — APPROVED** [review]

**Comment by notmd** [comment]
I have no idea why the tests only fail in CI, in local they still passing.

**Review by fozziethebeat — APPROVED** [review]
The web changes look good to me.  Does the local docker compose stack still work fine using these fake captcha credentials?

**Comment by fozziethebeat** [comment]
Let me copy your branch and see if i can replicate this test failure.  The fact that everything fails is worrisome

**Comment by notmd** [comment]
> The web changes look good to me. Does the local docker compose stack still work fine using these fake captcha credentials?

Yes, it still works.

> Let me copy your branch and see if i can replicate this test failure. The fact that everything fails is worrisome

All the tests failed because can't log in, which probably related to my changes

**Comment by notmd** [comment]
@fozziethebeat  I managed to reproduce the test fails in the local by running a full build (docker compose up). Will try to fix the tests tonight.

**Comment by notmd** [comment]
@fozziethebeat I have no idea to fix the tests so I disable them for now, lol. If you have time, please help me out.

**Comment by fozziethebeat** [comment]
@yk can you do a final review of the non-web changes?  They seem to address your request of an env variable to disable the captcha

**Review by andreaskoepf — APPROVED** [review]
.vscode default OK

**Review by yk — APPROVED** [review]
lgtm, thank you for the adjustments

**Comment by fozziethebeat** [comment]
Merging!

@notmd can you share in discord what env variables we have to setup on the deployment instances?  If you can generate them and send them to me privately I can add them to Github secrets

**Pull request merged**